libs/agno/agno/memory/v2/manager.py

```diff
@@ -44,30 +44,25 @@ def __init__(
         self.system_message = system_message
         self.memory_capture_instructions = memory_capture_instructions
         self.additional_instructions = additional_instructions
-
-    def add_tools_to_model(self, model: Model, tools: List[Callable]) -> None:
-        model = cast(Model, model)
-        model.reset_tools_and_functions()
-
-        _tools_for_model = []
-        _functions_for_model = {}
-
-        for tool in tools:
-            try:
-                function_name = tool.__name__
-                if function_name not in _functions_for_model:
-                    func = Function.from_callable(tool, strict=True)  # type: ignore
-                    func.strict = True
-                    _functions_for_model[func.name] = func
-                    _tools_for_model.append({"type": "function", "function": func.to_dict()})
-                    log_debug(f"Added function {func.name}")
-            except Exception as e:
-                log_warning(f"Could not add function {tool}: {e}")
-
-        # Set tools on the model
-        model.set_tools(tools=_tools_for_model)
-        # Set functions on the model
-        model.set_functions(functions=_functions_for_model)
+        self._tools_for_model: Optional[List[Dict[str, Any]]] = None
+        self._functions_for_model: Optional[Dict[str, Function]] = None
+
+    def determine_tools_for_model(self, tools: List[Callable]) -> None:
+        if self._tools_for_model is None:
+            self._tools_for_model = []
+            self._functions_for_model = {}
+
+            for tool in tools:
+                try:
+                    function_name = tool.__name__
+                    if function_name not in self._functions_for_model:
+                        func = Function.from_callable(tool, strict=True)  # type: ignore
+                        func.strict = True
+                        self._functions_for_model[func.name] = func
+                        self._tools_for_model.append({"type": "function", "function": func.to_dict()})
+                        log_debug(f"Added function {func.name}")
+                except Exception as e:
+                    log_warning(f"Could not add function {tool}: {e}")
```

Comment:
we still want the logic to run when `self._tools_for_model` already has tools, right? 
or is it cleaned on exit?

Reply:
else can we clarify what the early exit is about? it's a bit obscure rn

Reply:
Once it is populated we don't need it to run again.

Reply:
It saves on processing time